libcontainer/nsenter/idmap.h

```diff
@@ -0,0 +1,74 @@
+#include <sys/mount.h>
+// Centos-7 doesn't have this file nor the __has_include() directive, so let's
+// just leave this commented out and we can uncomment when it hits EOL (2024-06-30).
+//#include <linux/mount.h>
```

Comment:
So, I think this should be handled slightly differently. 

First, no need to `#include <linux/mount.h>`, as everything seems to be coming from other include files.

Second, maybe don't define syscall numbers if they don't exist on a platform used for compilation, instead doing something

```c
#if defined(MOUNT_ATTR_IDMAP) && defined(__NR_mount_setattr) && defined(__NR_open_tree)
#define HAVE_IDMAP_MOUNT
#else
#warning "not compiling in support for idmapped mounts"
#endif
```

and using `#ifdef HAVE_IDMAP_MOUNT` later in the code.

Reply:
I mean, have something like in a commit that adds the functionality
```c
void send_idmapsources(int sockfd, pid_t pid, char *idmap_src, int idmap_src_len)
#ifndef HAVE_IDMAP_MOUNT
{
    bail("no support for idmapped mounts");
}
#else
{
....
}
#endif /* HAVE_IDMAP_MOUNT */
```

Reply:
@kolyshkin regarding the first point, no, not everything is coming from other includes. Look at the mount setattr manpage: https://man7.org/linux/man-pages/man2/mount_setattr.2.html:

```
       #include <linux/mount.h> /* Definition of MOUNT_ATTR_* constants */
```

What helps us to drop that include and avoid hitting the issues the comment says is that we _do_ these defines here. So the define for `MOUNT_ATTR_IDMAP` comes from this file and we can drop the include to just fix the issues in old centos.

And centos-9 was also broken but in a more weird way (probably due to the patch to disable idmap mounts there, that is now being reverted) and I needed to add another ifdef here: https://github.com/opencontainers/runc/pull/3717/files#diff-421c257af6d93ef0bbd742757244adec93587030041c3306af559a1359e35532R27-R28 (centos-9 for some reason had __NR_mount_setattr defined but not the struct to pass the parameters, which is just so broken).

All of this to explain that supporting so many old OSes makes some things not ideal and it is not obvious if something like that will work. I'll try to see if there is a reasonable way to do something like you suggest. But this takes time, as the CI is slow and this runs in vagrant that is WAY more slower to run and painful to debug.

Reply:
@kolyshkin didn't manage to do anything good in this front.

I pushed what I was trying to this branch: https://github.com/kinvolk/runc/tree/rata/idmap-simpler-headers

And explained the problem there, but will put it here just in case:

 We can't avoid the defines of __NR_open_tree and __NR_mount_setattr.
 The thing is:
  * If we include linux/mount.h, then compilation on centos-7 fails (doesn't have that file). We can't
  use the __has_include() directive there as it seems the gcc version used
  doesn't support it.
  * So we need to not include linux/mount.h. But that means that we need to
  define the "struct mount_attr" and the constants: MOUNT_ATTR_IDMAP,
  OPEN_TREE_CLONE, OPEN_TREE_CLOEXEC. But sadly, even doing that is not
  enough, as then __NR_open_tree nor __NR_mount_setattr isn't defined either.

  So, basically, it seems we need to keep the defines here as long as we
  support centos-7.

Sounds good then to keep this as it is?

Reply:
Maybe there is something else that can be done for the first item (make centos-7 don't complain with that include), but I haven't found it.

Also, I was playing with this again today and I found another barrier to remove the defines of the syscalls. I don't think we can remove them anytime soon :(.

Ubuntu 20.04 ships with linux-libc-dev version 5.4.0-147.164 (or 5.4.whatever). This version was built for the 5.4 kernel headers, that was originally shipped with Ubuntu 20.04, and include open_tree but does include mount_sysattr. However, Ubuntu upgraded the kernel to 5.15 in this release (the github action uses a 5.15 kernel too, see this for the focal kernel package https://packages.ubuntu.com/search?keywords=linux-image-5.15.0-71-generic). It probably makes sense to keep the 5.4 headers for Ubuntu, so apps are unchanged, and just ship an updated kernel.

Therefore, we want to run on Ubuntu 20.04, but ubuntu ships with outdated headers. So we want to use our defines and make this feature just work on kernels that have the features we need.

I think this should solve the general concerns about the defines, as it really seems there is no way around it unless we drop support for lot of popular distros that are to be maintained several more years. I strongly think we should keep the defines now that we know this.

@kolyshkin what do you think?

Reply:
@kolyshkin Also, I added this: https://github.com/opencontainers/runc/pull/3717/files#diff-6383238247e090d88ade6343c0ef59dd09b3c10634bf0584e78445b843c55ab0R913-R914

I was already doing that for `mount_setattr(2)` but not for `open_tree(2)`. I think handling it this way has the same advantages than the version you proposed that just bailed.

So, hopefully this can be marked as solved @kolyshkin ? Please resolve if you agree :)

Reply:
Sounds good to me.